table/update_schema.go

```diff
@@ -0,0 +1,426 @@
+
+	return next
+}
+
+func (us *UpdateSchema) findField(path []string) *iceberg.NestedField {
+	name := strings.Join(path, ".")
+
+	var field iceberg.NestedField
+	var ok bool
+
+	if us.caseSensitive {
+		field, ok = us.schema.FindFieldByName(name)
+	} else {
+		field, ok = us.schema.FindFieldByNameCaseInsensitive(name)
+	}
+
+	if !ok {
+		return nil
+	}
+
+	return &field
+}
+
+func allowedPromotion(oldType, newType iceberg.Type) bool {
+
+	switch old := oldType.(type) {
+
+	case iceberg.PrimitiveType:
+		switch old.Type() {
+		case "int":
+			return newType.Type() == "long"
+		case "float":
+			return newType.Type() == "double"
+		case "decimal":
+			o := oldType.(iceberg.DecimalType)
+			n, ok := newType.(iceberg.DecimalType)
+			return ok && o.Scale() == n.Scale() && n.Precision() > o.Precision()
+		case "date":
+			return newType.Type() == "timestamp" || newType.Type() == "timestamp_ns"
+		}
+	}
+
+	return false
+}
+
+func (u *UpdateSchema) applyChanges() *iceberg.Schema {
+
+	newFields := rebuild(u.schema.AsStruct().FieldList, -1, u)
+
+	idList := u.schema.IdentifierFieldIDs
+	newID := u.schema.ID + 1
+
+	return iceberg.NewSchemaWithIdentifiers(newID, idList, newFields...)
+}
+
+func rebuild(fields []iceberg.NestedField, parentID int, u *UpdateSchema) []iceberg.NestedField {
+
+	var out []iceberg.NestedField
+
+	for _, f := range fields {
+		if _, gone := u.deletes[f.ID]; gone {
+			continue
+		}
+		if upd, ok := u.updates[f.ID]; ok {
+			f = *upd
+		}
+		switch t := f.Type.(type) {
+		case *iceberg.StructType:
+			f.Type = &iceberg.StructType{FieldList: rebuild(t.Fields(), f.ID, u)}
+		case *iceberg.ListType:
+			el := t.ElementField()
+			elSlice := rebuild([]iceberg.NestedField{el}, el.ID, u)
+			el = elSlice[0]
+			f.Type = &iceberg.ListType{
+				ElementID:       el.ID,
+				Element:         el.Type,
+				ElementRequired: el.Required,
+			}
+		case *iceberg.MapType:
+			val := t.ValueField()
+			valSlice := rebuild([]iceberg.NestedField{val}, val.ID, u)
+			val = valSlice[0]
+			f.Type = &iceberg.MapType{
+				KeyID:         t.KeyField().ID,
+				KeyType:       t.KeyField().Type,
+				ValueID:       val.ID,
+				ValueType:     val.Type,
+				ValueRequired: val.Required,
+			}
+		}
+		out = append(out, f)
+	}
+
+	// append new children for this parent
+	for _, nf := range u.adds[parentID] {
+		out = append(out, *nf)
+	}
+
+	return out
+}
+
+func validateDefaultValue(typ iceberg.Type, val any) error {
+
+	//Defaults are only allowed on primitive columns
+	prim, ok := typ.(iceberg.PrimitiveType)
+	if !ok {
+		return fmt.Errorf("defaults are only allowed on primitive columns, got %s", typ.Type())
+	}
+
+	lit, err := literalFromAny(val)
+	if err != nil {
+		return err
+	}
+
+	litType := lit.Type()
+
+	// Exact match?
+	if litType.Equals(prim) {
+		return nil
+	}
+
+	return fmt.Errorf("default literal of type %s is not assignable to of type %s", litType.String(), prim.Type())
+}
+
+func literalFromAny(v any) (iceberg.Literal, error) {
```

Comment:
this method looks like exact copy of `iceberg.NewLiteral`
why it's needed?

Reply:
was not able to call newLiteral directly like this here 
```
	lit := iceberg.NewLiteral(val)

```
was getting this error 

```
any does not satisfy iceberg.LiteralTypecompiler[InvalidTypeArg]
```

this was the only way i could call NewLiteral 
do let me know if there is a way 😓